rp2040-hal/src/clocks/macros.rs

```diff
@@ -0,0 +1,210 @@
+#[macro_use]
+
+macro_rules! int_division {
+    ($name:ident, $div:ident, $u:ty) => {
+        impl IntegerDivision for $name {
+            fn set_int_div(&mut self, div: usize) {
+                unsafe { self.shared_dev.get() }.$div.write(|w| unsafe {
+                    w.int().bits(div as $u);
+                    w
+                });
+            }
+        }
+    };
+}
+
+macro_rules! frac_division {
+    ($name:ident, $div:ident, $u:ty) => {
+        impl FractionDivision for $name {
+            fn set_frac_div(&mut self, div: usize) {
+                unsafe { self.shared_dev.get() }.$div.write(|w| unsafe {
+                    w.frac().bits(div as $u);
+                    w
+                });
+            }
+        }
+    };
+}
+
+macro_rules! clock_generator {
+    ($name:ident, $ctrl:ident) => {
+        impl ClockGenerator for $name {
+            fn enable(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
+                    w.enable().set_bit();
+                    w
+                });
+            }
+
+            fn disable(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
+                    w.enable().clear_bit();
+                    w
+                });
+            }
+
+            fn kill(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
+                    w.kill().set_bit();
+                    w
+                });
+            }
+        }
+    };
+}
+
+macro_rules! xosc_source {
+    ($name:ident, $ctrl:ident) => {
+        impl XOSCClockSource for $name {
+            fn set_xosc_src(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
+                    w.src().xosc_clksrc();
+                    w
+                });
+            }
+        }
+    };
+}
+
+macro_rules! rosc_source {
+    ($name:ident, $ctrl:ident) => {
+        impl ROSCClockSource for $name {
+            fn set_rosc_src(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
+                    w.src().rosc_clksrc_ph();
+                    w
+                });
+            }
+        }
+    };
+}
+
+macro_rules! selfaux_source {
+    ($name:ident, $ctrl:ident, $self:ident) => {
+        impl SelfAuxClockSource for $name {
+            fn set_self_aux_src(&mut self) {
+                unsafe { self.shared_dev.get() }.$ctrl.write(|w| {
```

Comment:
I mentioned above re. using `write` instead of `modify`, but here it completely prevents actually setting a different aux clock source besides what happens to be in `0x00`, making the `set_…_auxsrc` functions useless.